main/src/ca/uwaterloo/flix/language/phase/Weeder.scala

```diff
@@ -1464,6 +1469,59 @@ object Weeder extends Phase[ParsedAst.Program, WeededAst.Program] {
 
   }
 
+  /**
+    * Translates the hex code into the corresponding character.
+    * Returns an error if the code is not hexadecimal.
+    */
+  private def translateHexCode(code: String, loc: SourceLocation): Validation[Char, WeederError] = {
+    try {
+      Integer.parseInt(code, 16).toChar.toSuccess
+    } catch {
+      case _: NumberFormatException => WeederError.MalformedUnicodeEscape(code, loc).toFailure
+    }
+  }
+
+  /**
+    * Performs weeding on the given sequence of CharCodes.
+    */
+  private def weedCharSequence(chars0: Seq[ParsedAst.Literal.CharCode]): Validation[String, WeederError] = {
+
+    @tailrec
+    def visit(chars: List[ParsedAst.Literal.CharCode], acc: List[Char]): Validation[String, WeederError] = {
+      chars match {
+        case Nil => acc.reverse.mkString.toSuccess
+        case ParsedAst.CharCode.Literal(_, char, _) :: rest => visit(rest, char.head :: acc)
+        case ParsedAst.CharCode.Escape(_, "n", _) :: rest => visit(rest, '\n' :: acc)
+        case ParsedAst.CharCode.Escape(_, "r", _) :: rest => visit(rest, '\r' :: acc)
+        case ParsedAst.CharCode.Escape(_, "\\", _) :: rest => visit(rest, '\\' :: acc)
+        case ParsedAst.CharCode.Escape(_, "\"", _) :: rest => visit(rest, '\"' :: acc)
+        case ParsedAst.CharCode.Escape(_, "\'", _) :: rest => visit(rest, '\'' :: acc)
+        case ParsedAst.CharCode.Escape(_, "t", _) :: rest => visit(rest, '\t' :: acc)
```

Comment:
I would use a nested pattern match:

First match on Nil / Literal :: rest / Escape :: rest

Then for Escape you can have a sub pattern match on rest etc.